### PR TITLE
Add `is_finite` method for checking built-in types

### LINKED ISSUE #3790: Add a new 'is_finite' method for checking built-in types
### Describe the project you are working on

I am working on tools for the editor, 3D and 2D games. Of the programming languages I use only GDScript. I plan to host the tools in AssetLib, and it's important to me that my addons always handle exceptions. My addons have many Float-based properties open to the user, such as Vector2, Vector3, Transform and others.

### Describe the problem or limitation you are having in your project

As strange as it may sound, but Real numbers are not always really Real.
The problem lies in the lack of a universal way to check most of the built-in types using Real, such as Float, Transform (2D, 3D), Vector (2, 3), Rect2, Basis, AABB and others.
Yes, in @GDScript there are methods `is_inf(float)` and `is_nan(float)`, and in this case the record I need gets the following form: `not is_nan(value) and not is_inf(value)` - this takes up a lot of space. The entry for the other types looks a little longer, for example, `not is_nan(vector2.x + vector2.y) and not is_inf(vector2.x + vector2.y)`, for Vector3 it is not much longer, and for Transform it looks like this: `not is_nan((transform * Vector3()).length()) and not is_inf((transform * Vector3()).length())`
It's all very cumbersome. Prescribing the same method in different classes is not very convenient and contradicts the principles of OOP. The use of Singletons hinders decentralization and is often not considered by some programmers.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The solution to the problem is to add a new method to all the built-in types from the following list: Vector2, Vector3, Plane, Color, Rect2, Quat, AABB, Basis, Transform2D, Transform. And by adding a method to @GDScript for float. These methods will return true if all components of the built-in type fulfill the condition `not is_nan(component) and not is_inf(component)` and false in the opposite case.
This method can have the form as in many common programming languages: `bool is_finite(value: float)` and `bool type.is_finite()`

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```
print(is_finite(1.0)) # prints true
print(is_finite(NAN)) # prints false
print(is_finite(INF)) # prints false
print(Vector3(0.0, 0.0, INF).is_finite()) # prints false
var transform := Transform.IDENTITY
print(transform.is_finite()) # prints true
transform.origin.x = NAN
print(transform.is_finite()) # prints false
```
is_finite

### If this enhancement will not be used often, can it be worked around with a few lines of script?

If this is not used often, then it can be implemented in a few lines of script. But checking this exception may be required anywhere in any project.

### Is there a reason why this should be core and not an add-on in the asset library?

There is a reason, everything is the same as with the `is_inf` and `is_nan` methods. After all, they can be replaced with entries `value != INF and value != -INF` and `value != NAN`

### PR DESCRIPTION
**Pull request by timothyqiu**
Resolves https://github.com/godotengine/godot-proposals/issues/3790

This also makes fixing `NAN`/`INF` related crashes easier.

`is_finite` is not added for `Color` because a finite color sounds weird.

### CODE REVIEW

**Comment by dalexeev** [comment]
It makes sense to rename `is_inf` to `is_infinite`.

**Comment by timothyqiu** [comment]
> It makes sense to rename `is_inf` to `is_infinite`.

Yeah. But I think renaming it is out of the scope of this PR.

The naming is currently similar to C++ which provides both `isinf()` and `isinfinite()` :P

**Review by lawnjelly — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

File: core/math/basis.cpp [hunk 1]
```diff
@@ -691,6 +691,10 @@ bool Basis::is_equal_approx(const Basis &p_basis) const {
 	return rows[0].is_equal_approx(p_basis.rows[0]) && rows[1].is_equal_approx(p_basis.rows[1]) && rows[2].is_equal_approx(p_basis.rows[2]);
 }
 
+bool Basis::is_finite() const {
+	return rows[0].is_finite() && rows[1].is_finite() && rows[2].is_finite();
+}
```
Comment (aaronfranke) [inline core/math/basis.cpp:696]: Considering you can validate a Basis by checking its determinant, I'm not sure this is helpful. The same likely applies to several others. But I guess it's fine.

**Pull request merged**